example/eblob_backend.cpp

```diff
@@ -746,96 +749,269 @@ struct iterated_key_info {
+				return send_fail_reply(err);
 			}
+		}
 
-			auto response = serialize(ioremap::elliptics::dnet_iterator_response{
-				iterator_id, // iterator_id
-				info.key, // key
-				error.code(), // status
+		if (info->data_size <= request.chunk_size) {
+			if (info->data_size) {
+				data = data_pointer::allocate(info->data_size);
+				const int err = dnet_read_ll(info->fd, data.data<char>(), data.size(), info->data_offset);
+				if (err) {
+					dnet_backend_log(c->blog, DNET_LOG_ERROR,
+							 "EBLOB: server_send: %s: failed to read data: %s",
+							 dnet_dump_id_str(info->key.id), dnet_print_error(err));
+					return send_fail_reply(err);
+				}
+			}
 
-				counter, // iterated_keys
-				request.keys.size(), // total_keys
+			oplock_guard.unlock();
 
-				info.record_flags, // record_flags
-				info.ehdr.flags, // user_flags
+			monitor.add_bytes(info->jhdr.size + info->data_size);
 
-				info.jhdr.timestamp, // json_timestamp
-				info.jhdr.size, // json_size
-				info.jhdr.capacity, // json_capacity
-				0, // read_json_size
+			auto async = session->write(info->key,
+						    json, info->jhdr.capacity,
+						    data, info->data_size);
 
-				info.ehdr.timestamp, // data timestamp
-				info.data_size, // data_size
-				0, // read_data_size
-			});
+			async.connect([=, &request, &monitor] (const newapi::sync_write_result &/*results*/, const error_info &error) {
+					if (st->__need_exit) {
+						dnet_backend_log(c->blog, DNET_LOG_ERROR,
+								 "EBLOB: Interrupting server_send: peer has been disconnected");
+					} else {
+						auto response = serialize_response(error.code());
+						dnet_send_reply(st, cmd, response.data(), response.size(), 1);
+					}
 
-			dnet_send_reply(st, cmd, response.data(), response.size(), 1);
-		});
+					monitor.remove_bytes(info->jhdr.size + info->data_size);
+				});
+		} else {
+			uint64_t data_offset = 0;
+			data = data_pointer::allocate(request.chunk_size);
+
+			auto get_write_result = [&](newapi::async_lookup_result &async) -> int {
+				for (const auto &result: async.get()) {
+					const int status = result.status();
+					if (status != 0)
+						return status;
+				}
+				return 0;
+			};
+
+			while (data_offset < info->data_size && !st->__need_exit) {
+				const uint64_t remaining_size = info->data_size - data_offset;
+				const uint64_t data_size = std::min(remaining_size, request.chunk_size);
+				int err = dnet_read_ll(info->fd, data.data<char>(), data_size, data_offset);
+				if (err) {
+					dnet_backend_log(c->blog, DNET_LOG_ERROR,
+							 "EBLOB: server_send: %s: failed to read data: %s",
+							 dnet_dump_id_str(info->key.id), dnet_print_error(err));
+					return send_fail_reply(err);
+				}
+
+				if (data_offset == 0) {
+					auto async = session->write_prepare(info->key,
+									    json, info->jhdr.capacity,
+									    data, 0, info->data_size);
+					if ((err = get_write_result(async)) != 0) {
```

Comment:
`get_write_result` will lock current thread until result of write is received. Write by chunk should be done asynchronously without locking this thread.